File: linked_list/LC24.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    # 0,2,4 를 헤드로 자기 하나 뒤의것과 교환을 반복한다.
    def swapPairs(self, head: [ListNode]) -> [ListNode]:
        root = prev = ListNode(None)
        prev.next = head
        while head and head.next:
            b = head.next
            head.next = b.next
            b.next = head

            prev.next = b

            head = head.next
            prev = prev.next.next
        return root.next

    def swapPairs(self, head: [ListNode]) -> [ListNode]:
        if head and head.next:
            p = head.next
            head.next = self.swapPairs(p.next)
            p.next = head
            return p
        return head

File: linked_list/LC24.java
package linked_list;

public class LC24 {

    class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }

    }

    class Solution {
        public ListNode swapPairs(ListNode head) {

            ListNode prev = new ListNode();
            ListNode root = prev;
            prev.next = head;
            while (head != null && head.next != null) {
                ListNode tmp = head.next;
                head.next = tmp.next;
                tmp.next = head;
                prev.next = tmp;

                head = head.next;
                prev = prev.next.next;
            }

            return root.next;
        }
    }

}